geomstats/geometry/pullback_metric.py

```diff
@@ -103,6 +103,415 @@ def pickable_inner_product(i, j):
             pickable_inner_product(i, j)
             for i, j in itertools.product(range(self.dim), range(self.dim))
         )
-
         metric_mat = gs.reshape(gs.array(out), (-1, self.dim, self.dim))
         return metric_mat[0] if base_point.ndim == 1 else metric_mat
+
+
+class PullbackDiffeoMetric(RiemannianMetric, abc.ABC):
+    """
+    Pullback metric via a diffeomorphism.
+
+    Parameters
+    ----------
+    dim : int
+        Dimension.
+    """
+
+    def __init__(self, dim, shape):
```

Comment:
I agree ! I think this is more of a philosophical question and a choice of design. We ask ourselves the question with Yann when we worked on PullbackDiffeo, and here i the réflexion we came up with:

- function as argument is not always a good idea because of eventual border effects
- mathematically, in geomstats: abstractclass look to be referencing to category (an abstract kind of object that have a structure: https://www.wikiwand.com/en/Category_(mathematics)) (Vectorspace, levelset, openset, Riemanianmetric...) and instantiable class look to be referencing a specific space (Hypersphere, SpecialEuclidean, Poincarré...) those class still depend on some initialization parameter (dim, n...) but the structure is quite fixed. In this sense PullbackMetric are more of a category to me, and can be used in wide variety of structure cases. Thus, metaclass look to use more suited than instantiable class. Then, diffeo and inverse diffeo are part of the structure to be define and then are implemented as a method to be overwritten. I think this is very similar to what is done with basis in Vectorspace metaclass, it is given as a method, not an argument.
- Another point is that JacobianDiffeo and TangentDiffeo here have default implementation based on autodiff but can be overwritten when closed form are available as it is the case for Yann in some FullRankMatrice metrics. Then, we thought it is weird to give the diffeo as argument and its Jacobean as an overload method !

However, I do not fully know what are the design choices of @ninamiolane  and @nguigs